runtime/parachains/src/initializer/benchmarking.rs

```diff
@@ -0,0 +1,37 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+use super::*;
+use frame_benchmarking::{benchmarks, impl_benchmark_test_suite};
+use frame_system::{DigestItemOf, RawOrigin};
+use primitives::v1::ConsensusLog;
+use sp_runtime::traits::One;
+
+benchmarks! {
+	force_approve {}: _(RawOrigin::Root, One::one())
+	verify {
+		assert_eq!(
+			<frame_system::Pallet<T>>::digest().logs.last().unwrap(),
+			&<DigestItemOf<T>>::from(ConsensusLog::ForceApprove(One::one())),
+		);
+	}
+}
+
```

Comment:
Essentially I think the only storage op here is an `<Digest<T>>::append(item);` - without looking at the code I am guessing it has to decode the vec, push the item, and then encode it (But maybe it it is more optimized?). I am wondering if for something like this we should benchmark with a large vec to assume worst case?

My naive guess though is that practically we will never have enough logs for the append op to be that expensive + this is from a force origin so it doesn't look like a significant DoS vector.

Reply:
I looked through the code, and while it doesn't decode the entire vec (it only decodes the first 4 bytes to get the encoded length), it does however copy the entire vec to a new one whenever the new encoded length is different from the previous (which is pretty much always).

I'll most likely need to add in a variable and see how the complexity varies with the digest vec length.